### PR TITLE
#23 Add tests running inside a UDF

### LINKED ISSUE #23: Add integration test running in Exasol
To ensure that the driver runs inside an Exasol database we need to add integration tests.

Precondition: Exasol Docker container that includes dependencies.

### PR DESCRIPTION
**Pull request by kaklakariada**
Closes #23

Running tests requires a preview release of Exasol 8. See #78.

### CODE REVIEW

File: doc/design.md [hunk 1]
```diff
@@ -130,6 +130,28 @@ Covers:
 
 Needs: impl, utest
 
+## Restrict Libraries to the Ones Available to Exasol UDFs
```
Comment (redcatbear) [inline doc/design.md:133]: This is a constraint, not a requirement. So you should add a constraints section. See OFT templates for an example.

File: doc/system_requirements.md [hunk 2]
```diff
@@ -253,4 +253,22 @@ Covers:
 
 * [feat~logging~1](#logging)
 
-Needs: dsn
\ No newline at end of file
+Needs: dsn
+
+### Restrict Libraries to the Ones Available to Exasol UDFs
```
Comment (redcatbear) [inline doc/system_requirements.md:258]: Constraint, not requirement.

File: doc/user_guide/user_guide.md [hunk 3]
```diff
@@ -237,3 +237,49 @@ environment:close()
 ```
 
 Close the environment after you have closed all connections created with it. The `close()` method will return `false` if not all connections where closed or if the environment is already closed.
+
+## Using exasol-driver-lua in an Exasol UDF
```
Comment (redcatbear) [inline doc/user_guide/user_guide.md:241]:
```suggestion
## Using `exasol-driver-lua` in an Exasol UDF
```

File: doc/user_guide/user_guide.md [hunk 4]
```diff
@@ -237,3 +237,49 @@ environment:close()
...
+
+The Exasol in version 8 or later allows running Lua code in [user defined functions (UDF)](https://docs.exasol.com/db/latest/database_concepts/udf_scripts.htm). The exasol-driver-lua uses only dependencies that are available to UDFs or that can be included into an package using amalgamation. This makes it possible to also use it in an Exasol UDF, e.g. for accessing another Exasol database.
```
Comment (redcatbear) [inline doc/user_guide/user_guide.md:243]:
```suggestion
Exasol version 7.1 or later allows running Lua code in [user defined functions (UDF)](https://docs.exasol.com/db/latest/database_concepts/udf_scripts.htm). The exasol-driver-lua uses only dependencies that are available to UDFs or that can be included into an package using amalgamation. This makes it possible to also use it in an Exasol UDF, e.g. for accessing another Exasol database. Some required C-Lua-interface packages are shipped with Exasol 8 and later. So Exasol 8 is required to run the driver.
```

File: doc/user_guide/user_guide.md [hunk 5]
```diff
@@ -237,3 +237,49 @@ environment:close()
...
+    amalg.lua --fallback --script=udf.lua --output udf-amalg.lua \
+      luasql.exasol luasql.exasol.CursorData luasql.exasol.Environment \
+      luasql.exasol.Websocket luasql.exasol.WebsocketDatahandler \
+      luasql.exasol.Cursor luasql.exasol.util \
+      luasql.exasol.constants luasql.exasol.Connection \
+      luasql.exasol.ExasolWebsocket luasql.exasol.ConnectionProperties \
+      luasql.exasol.luws luasql.exasol.base64 \
+      remotelog exaerror message_expander
+    ```
+    This command adds all required modules of the driver as well as the third party modules `remotelog exaerror message_expander` to a single Lua file, using `udf.lua` as entry point.
```
Comment (redcatbear) [inline doc/user_guide/user_guide.md:270]:
```suggestion
    This command bundles all required modules of the driver as well as the third party modules `remotelog exaerror message_expander` to a single Lua file, using `udf.lua` as entry point.
```

File: spec/amalg_util.lua [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+local log = require("remotelog")
+
+local M = {}
+
+local function get_rockspec_filename() --
+    return string.format("luasql-exasol-%s.rockspec", constants.VERSION)
+end
+
+local function load_rockspec(path)
+    path = path or get_rockspec_filename()
```
Comment (redcatbear) [inline spec/amalg_util.lua:11]: Breaks SoC. Please move the default part logic outside of this function.

File: spec/amalg_util.lua [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    temp_file:close()
+    return path
+end
+
+local function get_lua_path()
+    return "src/?.lua"
+end
+
+local function get_third_party_module_names()
+    return {"remotelog", "exaerror", "message_expander"}
```
Comment (redcatbear) [inline spec/amalg_util.lua:56]: Extract constant. Move to beginning of the file.

File: spec/config.lua [hunk 8]
```diff
@@ -62,4 +61,33 @@ function M.create_connection()
...
+    if not M._exasol_version then
+        M._exasol_version = M._read_exasol_version()
+    end
+    return M._exasol_version
+end
+
+function M._read_exasol_version()
+    local conn = M.create_connection()
+    local cur = assert(conn:execute("select param_value from exa_metadata where param_name = 'databaseProductVersion'"))
+    local version = cur:fetch()[1]
```
Comment (redcatbear) [inline spec/config.lua:86]: If fetching fails, resources are not closed.

File: spec/integration/udf_spec.lua [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        assert.is_true(env:close())
+        env = nil
+    end)
+
+    local function create_schema()
+        local schema_name = string.format("CONNECTION_TEST_%d", os.time())
+        assert(conn:execute(string.format("drop schema if exists %s cascade", schema_name)))
+        assert(conn:execute(string.format("create schema %s", schema_name)))
+        finally(function()
+            assert(conn:execute(string.format("drop schema %s cascade", schema_name)))
```
Comment (redcatbear) [inline spec/integration/udf_spec.lua:32]:
```suggestion
            assert(conn:execute(string.format("drop schema if exists %s cascade", schema_name)))
```

File: spec/integration/udf_spec.lua [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        row = cur:fetch(row, "n")
+        index = index + 1
+    end
+    return result
+end
+
+function run(ctx)
+    local env = driver.exasol()
+    local conn = assert(env:connect(ctx.source_name, ctx.user_name, ctx.password))
+    local cur = assert(conn:execute(ctx.query))
```
Comment (redcatbear) [inline spec/integration/udf_spec.lua:70]: Connection will not be closed if anything past this point goes wrong.

File: spec/integration/udf_spec.lua [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    conn:close()
+    env:close()
+    return result
+end
+]]
+        local script_arguments =
+                "source_name VARCHAR(100), user_name VARCHAR(100), password VARCHAR(100), query VARCHAR(2000)"
+        local schema_name = create_schema()
+        create_script_with_driver(schema_name, script_arguments, script)
+        query = escape_string(query)
```
Comment (redcatbear) [inline spec/integration/udf_spec.lua:82]: Don't reuse variable. Just imagine it were immutable.

File: spec/integration/udf_spec.lua [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        cursor:close()
+        return result
+    end
+
+    it("can execute query in UDF", function()
+        if not config.db_supports_openssl_module() then
+            log.warn("Skipping test because current Exasol version does not support openssl module")
+            return
+        end
+        local result = execute_query_in_udf("select t.* from (values (1, 'a'), (2, 'b'), (3, 'c')) as t(num, txt)")
```
Comment (redcatbear) [inline spec/integration/udf_spec.lua:96]: :+1: Nice idea!

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: spec/amalg_util.lua (hunk 7 above)
Comment (kaklakariada) [reply spec/amalg_util.lua:56]: Done.

File: spec/integration/udf_spec.lua (hunk 11 above)
Comment (kaklakariada) [reply spec/integration/udf_spec.lua:82]: done.

File: spec/config.lua [hunk 13]
```diff
@@ -62,4 +61,68 @@ function M.create_connection()
...
+    local closeable = closable_creator()
+    return try_finally(function()
+        return try(closeable)
+    end, function()
+        closeable:close()
+    end)
+end
+
+function M._execute_query(query, fetchmode)
+    return try_with_closable(M.create_connection, function(conn)
```
Comment (redcatbear) [inline spec/config.lua:104]: :+1: Nice one!

**Review by redcatbear — APPROVED** [review]

**Pull request merged**